### ISSUE TITLE
Problem with tokens counter

### ISSUE THREAD

**Issue by pvtrn** [forestwanglin/openai-java#11]
when I count the number of tokens, this message is cut off after formatting and is not fully counted

```json
{
            "role": "tool",
            "content": "{\"items\":[{\"title\":\"#MINSK NIGHTLIFE TOUR / BELARUS AFTER SANCTIONS JUNE ...\", \"link\":\"https://www.youtube.com/watch?v=d6te-xePaj0\", \"snippet\":\"Jun 26, 2022 ... Anfisa BELARUS•66K views · 16:51 · Go to channel · MANHATTAN NIGHTLIFE AREAS - PACKED BARS \u0026 CLUBS Summer Update【ENTIRE TOUR】Best ...\"}, {\"title\":\"THE 10 BEST Nightlife Activities in Minsk (Updated 2024) - Tripadvisor\", \"link\":\"https://www.tripadvisor.com/Attractions-g294448-Activities-c20-Minsk.html\", \"snippet\":\"Results 1 - 30 of 77 ... These places are best for nightlife in Minsk: RetravelMe Belarus · Party Bus · HookahPlace Yakuba Kolasa · Nuahule Krasnaya · Dictator Bar.\"}, {\"title\":\"MINSK AT NIGHT WITH @IrishPartizan - YouTube\", \"link\":\"https://www.youtube.com/watch?v=1siveAZHym0\", \"snippet\":\"Aug 5, 2023 ... scene, this video is your ultimate guide to the hottest night spots in Minsk ... MINSK NIGHTLIFE AFTER SANCTIONS / BEST 5 NIGHT BARS IN MINSK.\"}]}",
            "name": "http",
            "tool_call_id": "call_ZPSUPvPgBhNMSzI2mXDW36XZ"
        }
```

```java
public static String formatArguments(String arguments) {
        List<String> lines = new ArrayList<>();
        lines.add("{");
        JSONObject jsonObject = JSONObject.parseObject(arguments);
        List<String> properties = new ArrayList<>();
        for (String fieldName : jsonObject.keySet()) {
            properties.add(String.format("\"%s\":%s", fieldName, formatValue(jsonObject.get(fieldName))));
        }
        lines.add(String.join(",\n", properties));
        lines.add("}");
        return String.join("\n", lines);
    }
```

Output: 
"{
"items":["","",""]
}"

**Comment by forestwanglin** [comment]
I tried to test your case as below. It won't go into the method `formatArguments`.  And the result I got seems to be correct.

```java 
    @Test
    public void issue11() {
        String json = """
                {
                            "role": "tool",
                            "content": "{\\"items\\":[{\\"title\\":\\"#MINSK NIGHTLIFE TOUR / BELARUS AFTER SANCTIONS JUNE ...\\", \\"link\\":\\"https://www.youtube.com/watch?v=d6te-xePaj0\\", \\"snippet\\":\\"Jun 26, 2022 ... Anfisa BELARUS•66K views · 16:51 · Go to channel · MANHATTAN NIGHTLIFE AREAS - PACKED BARS \\u0026 CLUBS Summer Update【ENTIRE TOUR】Best ...\\"}, {\\"title\\":\\"THE 10 BEST Nightlife Activities in Minsk (Updated 2024) - Tripadvisor\\", \\"link\\":\\"https://www.tripadvisor.com/Attractions-g294448-Activities-c20-Minsk.html\\", \\"snippet\\":\\"Results 1 - 30 of 77 ... These places are best for nightlife in Minsk: RetravelMe Belarus · Party Bus · HookahPlace Yakuba Kolasa · Nuahule Krasnaya · Dictator Bar.\\"}, {\\"title\\":\\"MINSK AT NIGHT WITH @IrishPartizan - YouTube\\", \\"link\\":\\"https://www.youtube.com/watch?v=1siveAZHym0\\", \\"snippet\\":\\"Aug 5, 2023 ... scene, this video is your ultimate guide to the hottest night spots in Minsk ... MINSK NIGHTLIFE AFTER SANCTIONS / BEST 5 NIGHT BARS IN MINSK.\\"}]}",
                            "name": "http",
                            "tool_call_id": "call_ZPSUPvPgBhNMSzI2mXDW36XZ"
                        }
                """;
        ChatMessage chatMessage = JSONObject.parseObject(json, ChatMessage.class);
        int tokenSize = TikTokenUtils.estimateTokensInMessage("gpt-4", chatMessage, 1);
        log.info("tokenSize: {}", tokenSize);
    }
```

I cannot reproduce your issue. Could you please paste all your code and let me know how you call the method?

**Comment by pvtrn** [comment]
```java
TikTokenUtils.estimateTokensInMessage("gpt-4", chatMessage, 2);
```

**Comment by forestwanglin** [comment]
Firstly, I don't know why you pass `2` as the third parameter.
Secondly, you can see the document in page [https://platform.openai.com/docs/api-reference/chat/create](https://platform.openai.com/docs/api-reference/chat/create).
The content must be a string type if role = `tool`.

I think maybe you invoke it in a wrong way.

You can ref the method `createFunctionChatCompletion`  in class [https://github.com/forestwanglin/openai-java/blob/main/service/src/test/java/xyz/felh/openai/OpenAiServiceTest.java](https://github.com/forestwanglin/openai-java/blob/main/service/src/test/java/xyz/felh/openai/OpenAiServiceTest.java)

**Comment by pvtrn** [comment]
This code count how many tool messages in chat. Therefore, this value is often greater than 1

```java
public static int estimateTokensInMessages(String modelName, List<ChatMessage> messages, List<Tool> tools) {
  int tokens = 0;
  int toolMessageSize = (int) messages.stream().filter(it -> it.getRole() == ChatMessageRole.TOOL).count(); // <== count 
  if (toolMessageSize > 1) {
      tokens += toolMessageSize * 2 + 1;

      int jsonContentToolSize = (int) messages.stream().filter(it -> it.getRole() == ChatMessageRole.TOOL
              && ToolContentFormat.isJSONString(it.getContent().toString())).count();
      if (jsonContentToolSize > 0) {
          tokens += 1 - jsonContentToolSize;
      }
  }
  boolean paddedSystem = false;
  for (ChatMessage message : messages) {
      ChatMessage msg = SerializationUtils.clone(message);
      if (msg.getRole() == ChatMessageRole.SYSTEM && Preconditions.isNotBlank(tools) && !paddedSystem) {
          if (Preconditions.isNotBlank(msg.getContent()) && msg.getContent() instanceof String) {
              msg.setContent(msg.getContent() + "\n");
          }
          paddedSystem = true;
      }
      tokens += estimateTokensInMessage(modelName, msg, toolMessageSize);
  }
  // Each completion (vs message) seems to carry a 3-token overhead
  tokens += 3;
  return tokens;
}
```

**Comment by forestwanglin** [comment]
My fault. I misunderstand your meaning. I've fixed this issue and deploy a new version `3.8.2024040101`.
